### PR TITLE
Add new ConversionKind to IsImplicitConversion test

### LINKED ISSUE #34984: ExtractMethodCodeRefactoringProvider throws InvalidOperationException on highlighting non-existent property
**Version Used**: 

**Steps to Reproduce**:

1. Highlight between $$

``` C#
// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.

using System;
using System.Collections.Generic;

namespace Microsoft.VisualStudio.ProjectSystem.VS.NuGet
{
    internal static class ImmutablePropertyCollectionEqualityComparer
    {
        public static bool AreUnorderedEqual<T>(ImmutablePropertyCollection<T> x, ImmutablePropertyCollection<T> y, IEqualityComparer<T> comparer)
            where T : class
        {
            Requires.NotNull(x, nameof(x));
            Requires.NotNull(y, nameof(y));

            if (x.Count != y.Count)
                return false;

            if (x.Count == 0)
                return true;

            foreach (T xItem in x.$$ToolReferences$$)
            {
                IVsReferenceItem yReference = y.ToolReferences.Item(xItem.Name);
                if (yReference == null)
                    return false;

                if (!ReferenceItemEqualityComparer.Instance.Equals(xReference, yReference))
                    return false;
            }
        }
    }
}

```
```
System.InvalidOperationException : Unexpected value 'UnsetConversionKind' of type 'Microsoft.CodeAnalysis.CSharp.ConversionKind'
   at Microsoft.CodeAnalysis.CSharp.ConversionKindExtensions.IsImplicitConversion(ConversionKind conversionKind)
   at Microsoft.CodeAnalysis.CSharp.Conversion.get_IsExplicit()
   at Microsoft.CodeAnalysis.CSharp.Utilities.SpeculationAnalyzer.ConversionsAreCompatible(Conversion originalConversion,Conversion newConversion)
   at Microsoft.CodeAnalysis.CSharp.Utilities.SpeculationAnalyzer.ForEachConversionsAreCompatible(SemanticModel originalModel,CommonForEachStatementSyntax originalForEach,SemanticModel newModel,CommonForEachStatementSyntax newForEach)
   at Microsoft.CodeAnalysis.Shared.Utilities.AbstractSpeculationAnalyzer`7.ReplacementBreaksForEachStatement(TForEachStatementSyntax forEachStatement,TForEachStatementSyntax newForEachStatement)
   at Microsoft.CodeAnalysis.Shared.Utilities.AbstractSpeculationAnalyzer`7.ReplacementChangesSemanticsForNode(SyntaxNode currentOriginalNode,SyntaxNode currentReplacedNode,SyntaxNode previousOriginalNode,SyntaxNode previousReplacedNode)
   at Microsoft.CodeAnalysis.Shared.Utilities.AbstractSpeculationAnalyzer`7.ReplacementChangesSemantics(SyntaxNode currentOriginalNode,SyntaxNode currentReplacedNode,SyntaxNode originalRoot,Boolean skipVerificationForCurrentNode)
   at Microsoft.CodeAnalysis.Shared.Utilities.AbstractSpeculationAnalyzer`7.ReplacementChangesSemantics()
   at Microsoft.CodeAnalysis.CSharp.Extensions.ExpressionSyntaxExtensions.CanReplaceWithReducedName(NameSyntax name,TypeSyntax reducedName,SemanticModel semanticModel,CancellationToken cancellationToken)
   at Microsoft.CodeAnalysis.CSharp.Extensions.ExpressionSyntaxExtensions.TryReduce(NameSyntax name,SemanticModel semanticModel,TypeSyntax& replacementNode,TextSpan& issueSpan,OptionSet optionSet,CancellationToken cancellationToken)
   at Microsoft.CodeAnalysis.CSharp.Extensions.ExpressionSyntaxExtensions.TryReduceExplicitName(ExpressionSyntax expression,SemanticModel semanticModel,TypeSyntax& replacementNode,TextSpan& issueSpan,OptionSet optionSet,CancellationToken cancellationToken)
   at Microsoft.CodeAnalysis.CSharp.Extensions.ExpressionSyntaxExtensions.TryReduceOrSimplifyExplicitName(ExpressionSyntax expression,SemanticModel semanticModel,ExpressionSyntax& replacementNode,TextSpan& issueSpan,OptionSet optionSet,CancellationToken cancellationToken)
   at Microsoft.CodeAnalysis.CSharp.Simplification.CSharpNameReducer.SimplifyName(SyntaxNode node,SemanticModel semanticModel,OptionSet optionSet,CancellationToken cancellationToken)
   at Microsoft.CodeAnalysis.CSharp.Simplification.AbstractCSharpReducer.AbstractReductionRewriter.SimplifyNode[TNode](TNode node,SyntaxNode newNode,SyntaxNode parentNode,Func`5 simplifier)
   at Microsoft.CodeAnalysis.CSharp.Simplification.AbstractCSharpReducer.AbstractReductionRewriter.SimplifyExpression[TExpression](TExpression expression,SyntaxNode newNode,Func`5 simplifier)
   at Microsoft.CodeAnalysis.CSharp.Simplification.CSharpNameReducer.Rewriter.VisitGenericName(GenericNameSyntax node)
   at Microsoft.CodeAnalysis.CSharp.Syntax.GenericNameSyntax.Accept[TResult](CSharpSyntaxVisitor`1 visitor)
   at Microsoft.CodeAnalysis.CSharp.CSharpSyntaxRewriter.Visit(SyntaxNode node)
   at Microsoft.CodeAnalysis.CSharp.Simplification.AbstractCSharpReducer.AbstractReductionRewriter.VisitInvocationExpression(InvocationExpressionSyntax node)
   at Microsoft.CodeAnalysis.CSharp.Syntax.InvocationExpressionSyntax.Accept[TResult](CSharpSyntaxVisitor`1 visitor)
   at Microsoft.CodeAnalysis.CSharp.CSharpSyntaxRewriter.Visit(SyntaxNode node)
   at Microsoft.CodeAnalysis.CSharp.CSharpSyntaxRewriter.VisitForEachStatement(ForEachStatementSyntax node)
   at Microsoft.CodeAnalysis.CSharp.Syntax.ForEachStatementSyntax.Accept[TResult](CSharpSyntaxVisitor`1 visitor)
   at Microsoft.CodeAnalysis.CSharp.CSharpSyntaxRewriter.Visit(SyntaxNode node)
   at Microsoft.CodeAnalysis.CSharp.Simplification.AbstractCSharpReducer.AbstractReductionRewriter.VisitNodeOrToken(SyntaxNodeOrToken nodeOrToken,SemanticModel semanticModel,Boolean simplifyAllDescendants)
   at async Microsoft.CodeAnalysis.Simplification.AbstractSimplificationService`3.<>c__DisplayClass10_1.<ReduceAsync>b__0[TExpressionSyntax,TStatementSyntax,TCrefSyntax](<Unknown Parameters>)
   at async Microsoft.CodeAnalysis.Simplification.AbstractSimplificationService`3.ReduceAsyncInternal[TExpressionSyntax,TStatementSyntax,TCrefSyntax](<Unknown Parameters>)
   at async Microsoft.CodeAnalysis.Simplification.AbstractSimplificationService`3.ReduceAsync[TExpressionSyntax,TStatementSyntax,TCrefSyntax](<Unknown Parameters>)
   at async Microsoft.CodeAnalysis.Simplification.Simplifier.ReduceAsync(<Unknown Parameters>)
   at async Microsoft.CodeAnalysis.ExtractMethod.MethodExtractor.CodeGenerator`3.GenerateAsync[TStatement,TExpression,TNodeUnderContainer](<Unknown Parameters>)
   at async Microsoft.CodeAnalysis.ExtractMethod.MethodExtractor.ExtractMethodAsync(<Unknown Parameters>)
   at async Microsoft.CodeAnalysis.ExtractMethod.AbstractExtractMethodService`3.ExtractMethodAsync[TValidator,TExtractor,TResult](<Unknown Parameters>)
   at async Microsoft.CodeAnalysis.CodeRefactorings.ExtractMethod.ExtractMethodCodeRefactoringProvider.GetCodeActionAsync(<Unknown Parameters>)
   at async Microsoft.CodeAnalysis.CodeRefactorings.ExtractMethod.ExtractMethodCodeRefactoringProvider.ComputeRefactoringsAsync(<Unknown Parameters>)
   at async Microsoft.CodeAnalysis.CodeRefactorings.CodeRefactoringService.GetRefactoringFromProviderAsync(<Unknown Parameters>)
   at Microsoft.VisualStudio.Telemetry.WindowsErrorReporting.WatsonReport.GetClrWatsonExceptionInfo(Exception exceptionObject)
```

### PR DESCRIPTION
**Pull request by ryzngard**
Fixes #34984 

`UnsetConversionKind` was added by [this commit](https://github.com/dotnet/roslyn/commit/9160b423#diff-61c14d55196186c0f71024c92e5a382fR11) but the switch statement wasn't updated.

### CODE REVIEW

**Comment by ryzngard** [comment]
Is there a place I can add tests? Just adding this code to our CodeAction tests didn't trigger the bug, but likely somewhere in the compiler layer there's a good place to add one.

**Comment by jcouv** [comment]
@ryzngard For testing, based on what I can see in the issue (stack trace), you need to get the conversion that is causing problems (likely from a semantic model call) and check `IsExplicit` on it. You need to debug the IDE code to figure out where the conversion came from (what syntax, regular or speculative semantic model, ...).

In terms of where to add the new compiler test(s), I see that `Conversion.IsExplicit` seems particularly covered in `GetSemanticInfoTests.cs`. Seems a good place.

**Comment by jcouv** [comment]
Marked PR as "private" for now. Please ping if you have any other questions, or once ready for review after test is added. Thanks

**Comment by ryzngard** [comment]
@jcouv  test added to cover this bug, should be ready for review

File: src/Compilers/CSharp/Test/Symbol/Compilation/GetSemanticInfoTests.cs [hunk 1]
```diff
@@ -5916,5 +5916,36 @@ public void GetSpecialType_ThrowsOnGreaterThanCount()
...
+        public async Task ConversionIsExplicit_UnsetConversionKind()
+        {
+            var source =
+@"class C1
+{
+}
+
+class C2
+{
+    public void Foo()
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/Symbol/Compilation/GetSemanticInfoTests.cs]: >Foo [](start = 16, length = 3)

nit: Foo is disallowed by policheck

File: src/Compilers/CSharp/Test/Symbol/Compilation/GetSemanticInfoTests.cs [hunk 2]
```diff
@@ -5916,5 +5916,36 @@ public void GetSpecialType_ThrowsOnGreaterThanCount()
...
+        var c = new C1();
+        foreach (string item in c.Items)
+        {
+        }
+}";
+            var comp = CreateCompilation(source);
+            var tree = comp.SyntaxTrees.Single();
+            var model = comp.GetSemanticModel(tree);
+
+            var root = await tree.GetRootAsync();
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/Symbol/Compilation/GetSemanticInfoTests.cs]: >await tree.GetRootAsync(); [](start = 23, length = 26)

No need to use async test method. Can just use `var root = tree.GetRoot();` here

File: src/Compilers/CSharp/Test/Symbol/Compilation/GetSemanticInfoTests.cs [hunk 3]
```diff
@@ -5916,5 +5916,36 @@ public void GetSpecialType_ThrowsOnGreaterThanCount()
...
+            var foreachSyntaxNode = root.DescendantNodes().OfType<ForEachStatementSyntax>().Single();
+            var foreachSymbolInfo = model.GetForEachStatementInfo(foreachSyntaxNode);
+
+            Assert.Equal(Conversion.UnsetConversion, foreachSymbolInfo.CurrentConversion);
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/Symbol/Compilation/GetSemanticInfoTests.cs]: >UnsetConversion [](start = 36, length = 15)

Feels like there is a bigger issue here. I'd expect the conversion to be `Conversion.NoConversion` (and `Exists` to be `false`).

I think that was changed by PR [https://github.com/dotnet/roslyn/pull/33648](https://github.com/dotnet/roslyn/pull/33648) which introduced `UnsetConversion`, but was not intentional.
Let's discuss with @gafter

File: src/Compilers/CSharp/Test/Symbol/Compilation/GetSemanticInfoTests.cs (hunk 1 above)
Comment (RikkiGibson) [reply src/Compilers/CSharp/Test/Symbol/Compilation/GetSemanticInfoTests.cs]: I suggest the name `M` :)

File: src/Compilers/CSharp/Test/Symbol/Compilation/GetSemanticInfoTests.cs (hunk 3 above)
Comment (ryzngard) [reply src/Compilers/CSharp/Test/Symbol/Compilation/GetSemanticInfoTests.cs]: I'm going to continue with this PR as is to fix the crash. I've filed #35918 to track a deeper fix if needed.

File: src/Compilers/CSharp/Test/Symbol/Compilation/GetSemanticInfoTests.cs (hunk 1 above)
Comment (ryzngard) [reply src/Compilers/CSharp/Test/Symbol/Compilation/GetSemanticInfoTests.cs]: `M` sounds good to me

**Comment by ryzngard** [comment]
@jcouv  @RikkiGibson  please take another look at this. I've opened a separate issue for the open concern #35918

**Comment by gafter** [comment]
@jcouv and @agocke are out of the office right now so it may take a little while to complete this review.

**Comment by gafter** [comment]
@ryzngard The compiler team requires two reviews before merging.  Generally speaking you should let each team merge reviews for code maintained by that team.  Hopefully @jcouv will be able to look at this on Tuesday.  I am not going to revert this, but it is possible there will be more work on the basis of his review.

**Comment by ryzngard** [comment]
@gafter my apologies. I'll remember that for the future.

**Pull request merged**